Sources/SourceKitLSP/Swift/SemanticTokens.swift

```diff
@@ -15,8 +15,33 @@ import LanguageServerProtocol
 import SwiftIDEUtils
 import SwiftParser
 import SwiftSyntax
+import SourceKitD
 
 extension SwiftLanguageServer {
+  /// Requests the semantic highlighting tokens for the given snapshot from sourcekitd.
+  private func semanticHighlightingTokens(for snapshot: DocumentSnapshot) async throws -> [SyntaxHighlightingToken]? {
+    guard let buildSettings = await self.buildSettings(for: snapshot.uri), !buildSettings.isFallback else {
+      return nil
+    }
+    
+    let skreq = SKDRequestDictionary(sourcekitd: self.sourcekitd)
+    skreq[keys.request] = requests.semantic_tokens
+    skreq[keys.sourcefile] = snapshot.uri.pseudoPath
+
+    // FIXME: SourceKit should probably cache this for us.
+    skreq[keys.compilerargs] = buildSettings.compilerArgs
+
+    let dict = try await sourcekitd.send(skreq)
+
+    guard let skTokens: SKDResponseArray = dict[keys.semantic_tokens] else {
+      return nil
+    }
+    let tokenParser = SyntaxHighlightingTokenParser(sourcekitd: sourcekitd)
+    var tokens: [SyntaxHighlightingToken] = []
+    tokenParser.parseTokens(skTokens, in: snapshot, into: &tokens)
```

Comment:
I assume there was (maybe still is?) some reason for this, but could `parseTokens` just return an array instead?

Reply:
`parseTokens` calls recursively into itself for substructures. and it’s easier + more efficient to accumulate tokens in an `inout Array`. At least, that’s what I think what the reasoning was.

Reply:
That could be internal to `SyntaxHighlightingToken`, I don't see why callers have to care about that.

Reply:
Since there’s only one caller, I didn’t care too much. But I made the `inout` functions private and added an `internal` entry point that returns the tokens.